source/fem-dem/vans_assemblers.cc

```diff
@@ -0,0 +1,607 @@
+              const Tensor<1, dim> &phi_u_j = scratch_data.phi_u[q][j];
+
+              local_matrix(i, j) +=
+                void_fraction * phi_u_j * phi_u_i * bdf_coefs[0] * JxW[q];
+            }
+        }
+    }
+}
+
+template <int dim>
+void
+GLSVansAssemblerBDF<dim>::assemble_rhs(
+  NavierStokesScratchData<dim> &        scratch_data,
+  StabilizedMethodsTensorCopyData<dim> &copy_data)
+{
+  // Loop and quadrature informations
+  const auto &       JxW        = scratch_data.JxW;
+  const unsigned int n_q_points = scratch_data.n_q_points;
+  const unsigned int n_dofs     = scratch_data.n_dofs;
+
+  // Copy data elements
+  auto &strong_residual = copy_data.strong_residual;
+  auto &local_rhs       = copy_data.local_rhs;
+
+  // Time stepping information
+  const auto          method = this->simulation_control->get_assembly_method();
+  std::vector<double> time_steps_vector =
+    this->simulation_control->get_time_steps_vector();
+
+  // Vector for the BDF coefficients
+  Vector<double> bdf_coefs = bdf_coefficients(method, time_steps_vector);
+  std::vector<Tensor<1, dim>> velocity(1 +
+                                       number_of_previous_solutions(method));
+  std::vector<double> void_fraction(1 + number_of_previous_solutions(method));
+
+  // Loop over the quadrature points
+  for (unsigned int q = 0; q < n_q_points; ++q)
+    {
+      velocity[0] = scratch_data.velocity_values[q];
+      for (unsigned int p = 0; p < number_of_previous_solutions(method); ++p)
+        velocity[p + 1] = scratch_data.previous_velocity_values[p][q];
+
+      void_fraction[0] = scratch_data.void_fraction_values[q];
+      for (unsigned int p = 0; p < number_of_previous_solutions(method); ++p)
+        void_fraction[p + 1] = scratch_data.previous_void_fraction_values[p][q];
+
+      for (unsigned int p = 0; p < number_of_previous_solutions(method) + 1;
+           ++p)
+        {
+          strong_residual[q] += bdf_coefs[p] * velocity[p] * void_fraction[0];
+        }
+
+
+      for (unsigned int i = 0; i < n_dofs; ++i)
+        {
+          const auto phi_u_i     = scratch_data.phi_u[q][i];
+          const auto div_phi_u_i = scratch_data.div_phi_u[q][i];
+          double     local_rhs_i = 0;
+          for (unsigned int p = 0; p < number_of_previous_solutions(method) + 1;
+               ++p)
+            {
+              local_rhs_i -=
+                (bdf_coefs[p] * void_fraction[0] * (velocity[p] * phi_u_i) +
+                 bdf_coefs[p] * (void_fraction[p] * scratch_data.phi_p[q][i]));
+
+              if (cfd_dem.grad_div == true)
+                {
+                  local_rhs_i -=
+                    (bdf_coefs[p] * void_fraction[p]) * div_phi_u_i;
+                }
+            }
+          local_rhs(i) += local_rhs_i * JxW[q];
+        }
+    }
+}
+
+template class GLSVansAssemblerBDF<2>;
+template class GLSVansAssemblerBDF<3>;
+
+
+template <int dim>
+void
+GLSVansAssemblerPFI<dim>::calculate_particle_fluid_interactions(
```

Comment:
So you made a single assembler for all of the particle_fluid_interaction. This is ok for now, but I think it would be best to eventually move to one assembler for each particle fluid interaction term.
One for each drag model, one for the pressure force, etc. You can do this in a next PR since this is already a big PR.